flight_computer/lib/opt3001.py

```diff
@@ -0,0 +1,156 @@
+        # specifically, low power is M[1:0] set to 0b00,
+        # but we can just reset everything
+        self.buf[1] = 0x00
+        self.buf[2] = 0x00
+        self.write_value(CONFIGURATION)
+
+        # TODO: is there anything else we should do? is the buffer automatically deallocated?
+
+    # CONFIGURE sun sensor on startup
+    # sets conversion time equal to chosen value (800ms or 100ms)
+    # currently automatic range setting & continuous conversions
+    def configure(self, conversion_time=DEFAULT_CONVERSION_TIME):
+        # automatic range setting and continuous conversions.
+
+        # configuration register fields:
+        # RN3 RN2 RN1 RN0 CT  M1 M0  OVF
+        # CRF FH  FL  L   POL ME FC1 FC0
+
+        # automatically set range RN[3:0] for best results
+        # Conversion time (CT) 800ms is 1, 100ms is 0
+        # to set continuous conversions: M[1:0] is set to 0b11
+        self.buf[1] = (AUTO_RANGE_SETTING << 4) \
+                      + ((conversion_time == MAX_CONVERSION_TIME) << 3) \
+                      + CONTINUOUS_CONVERSIONS << 1
+        self.buf[2] = 0x00
+        self.write_value(CONFIGURATION)
+
+        # TODO: Check to make sure all other values in both buf[1] and buf[2] are desirable
+
+        # TODO: Wait for conversion ready flag to be set
+
+    # READ value into last two bytes
+    def read_value(self, register):
+        # 'with' handles locking and unlocking the i2c
+        with self.i2c as i2c:
+            self.buf[0] = register
+            i2c.write_then_readinto(self.buf, self.buf, out_end=1, in_start=1)
+            # partial write so that we are reading from correct register field, (out_end = 1)
+            # should automatically be generating a start bit before reading,
+            # which is all we need (sending a stop bit is also fine according to opt3001 docs)
+            # then read but don't overwrite current register field (in_start = 1)
+            i2c.stop()
+
+    # WRITE value from last two bytes
+    def write_value(self, register):
+        # 'with' handles locking and unlocking the i2c
+        with self.i2c as i2c:
+            self.buf[0] = register
+            i2c.write(self.buf)
+            # automatically writes a stop bit
+
+    # CHECK whether sensor measurement is ready
+    @property
+    def conversion_ready(self):
+        # configuration register fields:
+        # RN3 RN2 RN1 RN0 CT  M1 M0  OVF
+        # CRF FH  FL  L   POL ME FC1 FC0
+
+        self.read_value(self, CONFIGURATION)
+        # CRF (conversion ready flag, see registers below)
+        # first most significant bit (last most position)
+        # in the second value byte equals 1 when ready
+        return extract_bits(self.buf[2], len(self.buf[2]) - 1, 1)
+
+    # LUX value of sun-sensor
+    # could have delay of 100ms or 800ms
+    @property
+    def lux(self):
+        # wait until the lux measurement is ready
+        while not self.conversion_ready:
+            time.sleep(0.1)
```

Comment:
It might be better not to sleep at all here, and just pass (this is called just spin waiting)